docs/closure-compiler.md

```diff
@@ -0,0 +1,697 @@
+};
+
+window.settings = SETTINGS;
+
+/** @const {!Object<string, string>} **/
+const DEACTIVATION_ACTIVATION_PAIRS = {
+  'mouseup': 'mousedown',
+  'pointerup': 'pointerdown',
+  'touchend': 'touchstart',
+  'keyup': 'keydown',
+  'blur': 'focus',
+};
+
+Object.keys(DEACTIVATION_ACTIVATION_PAIRS).forEach((deactivationEvt) => {
+  const activationEvt = DEACTIVATION_ACTIVATION_PAIRS[deactivationEvt];
+  domNode.addEventListener(activationEvt, someActivationListener);
+  domNode.addEventListener(deactivationEvt, someDeactivationListener);
+});
+
+/** @const {!Array<!Object<string, !Function>>} */
+const listeners = [
+  {
+    'mouseup': () => console.log('mouseup'),
+    'mousedown': () => console.log('mousedown'),
+  },
+  {
+    'keyup': () => console.log('keyup'),
+    'keydown': () => console.log('keydown'),
+  },
+];
+
+console.log(listeners[0]['mouseup']);
+```
+
+Sometimes in our code, object keys will have meaning outside just being a key for an object. An
+example of this might be a map of event types to their respective listeners, as shown above.
+
+When closure compiles javascript using advanced optimizations, it [obfuscates property values](https://developers.google.com/closure/compiler/docs/api-tutorial3#better) in
+objects, classes, etc. While this leads to smaller code, it also leads to issues when those
+properties are used externally. In order to prevent this behavior, object keys need to be quoted,
+and those keys need to be referenced using bracket notation.
+
+### Member declarations via sentinel property reference expressions.
+
+Closure enforces that properties cannot be added to object instances unless those properties are
+specified in their constructors. This is problematic in cases where a base constructor calls a
+"setup" function in which properties are added to the instance. This is a pattern we use in our
+codebase via `MDCComponent#initialize()`, so that the component has an opportunity to perform any
+instantiation logic without losing all of the base constructor logic of assigning a root element,
+instantiating an adapter, etc. Because `initialize()` is called as part of the `super()` call within
+the constructor, we cannot assign a default value to `this.innerEl_` within the constructor since it
+would override what's been assigned in `initialize()`. Furthermore, closure mandates that `super()`
+be the first expression within a method. Thus, we cannot assign to `this.innerEl_` before `super()`
+is called.
+
+To get around this, we simply create a sentinel expression statement that references the property.
+This sentinel expression lets closure know of the property declaration, which is declared _within_
+the constructor via a different function.
+
+**Example:**
+
+```js
+/** @final @extends {MDCComponent<!MyComponentFoundation>} */
+class MyComponent extends MDCComponent {
+  /**
+   * @param {...?} args
+   */
+  constructor(...args) {
+    super(...args);
+    /** @private {!Element} */
+    this.innerEl_;
```

Comment:
nit: add comment to make it more explicit this is the sentinel expression statement we are adding.